improver/calibration/reliability_calibration.py

```diff
@@ -327,15 +326,83 @@ def _populate_reliability_bins(self, forecast, truth):
             forecast_probabilities.append(forecasts_probability_values)
             forecast_counts.append(forecast_mask)
 
-        reliability_table = np.stack(
+        reliability_table = np.ma.stack(
             [
-                np.stack(observation_counts),
-                np.stack(forecast_probabilities),
-                np.stack(forecast_counts),
+                np.ma.stack(observation_counts),
+                np.ma.stack(forecast_probabilities),
+                np.ma.stack(forecast_counts),
             ]
         )
+
         return reliability_table.astype(np.float32)
 
+    def _populate_masked_reliability_bins(self, forecast, truth):
+        """
+        Support populating the reliability table bins with a masked truth. If a
+        masked truth is provided, a masked reliability table is returned.
+
+        Args:
+            forecast (numpy.ndarray):
+                An array containing data over an xy slice for a single validity
+                time and threshold.
+            truth (numpy.ma.MaskedArray):
+                An array containing a thresholded gridded truth at an
+                equivalent validity time to the forecast array.
+        Returns:
+            numpy.ma.MaskedArray:
+                An array containing reliability table data for a single time
+                and threshold. The leading dimension corresponds to the rows
+                of a calibration table, the second dimension to the number of
+                probability bins, and the trailing dimensions are the spatial
+                dimensions of the forecast and truth cubes (which are
+                equivalent).
+        """
+        forecast = np.ma.masked_where(np.ma.getmask(truth), forecast)
+        table = self._populate_reliability_bins(forecast, truth)
+        table.data[table.mask] = 0
```

Comment:
The data underneath the masks are zeroed because of the way that the reliability tables are added when a masked truth is present:
```python
            threshold_reliability = np.ma.array(
                threshold_reliability.data + table.data, mask=mask, dtype=np.float32,
            )
```
In this case, `threshold_reliability` and `table` are numpy arrays, so calling the data attribute is looking underneath the mask. The addition of the unmasked reliability tables is then masked using the bitwise addition of the masks. This means that only points that are masked in both reliability tables are kept as masked, which is not the usual behaviour when adding two masked arrays, as usually the mask would tend to grow as the addition of a masked and unmasked point will result in a masked point. This [StackOverflow question](https://stackoverflow.com/questions/20539069/add-together-two-numpy-masked-arrays) gives the general idea. By adding the masked arrays in this way, we're maximising the number of points that can contribute to the reliability tables, rather than ignoring points that are potentially only masked within one timestep of the training dataset.

I've added a comment to explain the zeroing of the data underneath the mask.

Reply:
Okay, thanks. That is clear now.